src/crypto/PrivateKeyBundle.ts

```
@@ -1,11 +1,124 @@
+    })
+  }
+
+  public getLinkedAddress(role: AccountLinkedRole): string {
+    return this.accountLinkedKey.publicKey.getLinkedAddress(role)
+  }
+
+  // Return a key bundle with the current pre-key.
+  public getPublicKeyBundle(): SignedPublicKeyBundleV2 {
+    if (!this._publicKeyBundle) {
+      this._publicKeyBundle = new SignedPublicKeyBundleV2({
+        accountLinkedKey: this.accountLinkedKey.publicKey,
+        preKey: this.getCurrentPreKey().publicKey,
+      })
+    }
+    return this._publicKeyBundle
+  }
+
+  // Return the current (latest) pre-key (to be advertised).
+  public getCurrentPreKey(): SignedPrivateKey {
+    return this.preKeys[0]
+  }
+
+  // Find pre-key matching the provided public key.
+  private findPreKey(which: SignedPublicKey): SignedPrivateKey {
+    const preKey = this.preKeys.find((key) => key.matches(which))
+    if (!preKey) {
+      throw new NoMatchingPreKeyError(which)
+    }
+    return preKey
+  }
+
+  // sharedSecret derives a secret from peer's key bundles using a variation of X3DH protocol
+  // where the sender's ephemeral key pair is replaced by the sender's pre-key.
+  // @peer is the peer's public key bundle
+  // @myPreKey indicates which of my preKeys should be used to derive the secret
+  // @recipient indicates if this is the sending or receiving side.
+  public async sharedSecret(
+    peer: SignedPublicKeyBundleV2,
+    myPreKey: SignedPublicKey,
+    isRecipient: boolean
+  ): Promise<Uint8Array> {
+    if (!peer.accountLinkedKey || !peer.preKey) {
+      throw new Error('invalid peer key bundle')
+    }
+    if (!(await peer.accountLinkedKey.verifyKey(peer.preKey))) {
+      throw new Error('peer preKey signature invalid')
+    }
+    if (!this.accountLinkedKey) {
+      throw new Error('missing account linked key')
+    }
+    let dh1: Uint8Array, dh2: Uint8Array, preKey: SignedPrivateKey
+    if (isRecipient) {
+      if (
+        !this.getLinkedAddress(AccountLinkedRole.INBOX_KEY) ||
```

Sorry for broken record, but let's double check this will fail rather than returning a garbage address. I think xmtp-js currently allows "garbage" addresses in this method, but the address will get surfaced to the user as an unexpected one.